pkg/proxy/index/index_controller.go

```diff
@@ -0,0 +1,294 @@
+		},
+		DeleteFunc: func(obj interface{}) {
+			if final, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+				obj = final.Obj
+			}
+			shard := obj.(*tenancyv1alpha1.ClusterWorkspaceShard)
+
+			c.lock.Lock()
+			defer c.lock.Unlock()
+			delete(c.shardBaseURLs, shard.Name)
+
+			c.enqueueShard(shard)
+		},
+	})
+
+	return c, nil
+}
+
+// Controller watches ClusterWorkspaceShards on the root shard, and then starts informers
+// for every ClusterWorkspaceShards, watching the ClusterWorkspaces on them. It then
+// updates the workspace index, which maps logical clusters to shard URLs.
+type Controller struct {
+	queue workqueue.RateLimitingInterface
+
+	rootHost     string
+	clientGetter ClusterWorkspaceClientGetter
+
+	clusterWorkspaceShardIndexer cache.Indexer
+	clusterWorkspaceShardLister  tenancylister.ClusterWorkspaceShardLister
+
+	clusterWorkspaceHandler cache.ResourceEventHandler
+
+	shardInformersLock             sync.RWMutex
+	shardClusterWorkspaceInformers map[string]cache.SharedIndexInformer
+	shardClusterWorkspaceStopCh    map[string]chan struct{}
+
+	lock                sync.RWMutex
+	workspaceShardNames map[logicalcluster.Name]string
+	shardBaseURLs       map[string]string
+}
+
+// Start the controller. It does not really do anything, but to keep the shape of a normal
+// controller, we keep it.
+func (c *Controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer func() {
+		c.shardInformersLock.Lock()
+		defer c.shardInformersLock.Unlock()
+		for _, stopCh := range c.shardClusterWorkspaceStopCh {
+			close(stopCh)
+		}
+	}()
+
+	klog.Infof("Starting %s controller", controllerName)
+	defer klog.Infof("Shutting %s controller", controllerName)
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *Controller) enqueueShard(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	klog.Infof("Enqueueing ClusterWorkspaceShard %q", key)
+
+	c.queue.Add(key)
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *Controller) process(ctx context.Context, key string) error {
+	shard, err := c.clusterWorkspaceShardLister.Get(key) // TODO: clients need a way to scope down the lister per-cluster
```

Comment:
will `shards` always be on the same kcp instance?

Reply:
yes